common/src/main/java/com/wynntils/models/players/GuildModel.java

```diff
@@ -4,54 +4,108 @@
 import com.wynntils.models.players.type.GuildRank;
 import com.wynntils.utils.mc.LoreUtils;
+import com.wynntils.utils.mc.McUtils;
 import java.util.List;
+import java.util.Locale;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
 
 public class GuildModel extends Model {
-    private static final Pattern GUILD_NAME_MATCHER = Pattern.compile("§3(.*?)§b.*");
-    private static final Pattern GUILD_RANK_MATCHER = Pattern.compile("§7Rank: §f(.*)");
+    // Test suite: https://regexr.com/7hob9
+    private static final Pattern GUILD_NAME_MATCHER = Pattern.compile("§3([a-zA-Z ]*?)§b \\[[a-zA-Z]{3,4}]");
 
-    private String name = "";
-    private GuildRank rank;
+    // Test suite: https://regexr.com/7hobf
+    // Recruiter is intentionally first, otherwise Recruit will incorrectly match first
+    // This applies to all other patterns in this class involving guild ranks
+    private static final Pattern GUILD_RANK_MATCHER =
+            Pattern.compile("§7Rank: §f(Recruiter|Recruit|Captain|Strategist|Chief|Owner)");
+
+    // Test suite: https://regexr.com/7hoae
+    private static final Pattern MSG_LEFT_GUILD = Pattern.compile("§3You have left §b[a-zA-Z ]*§3!");
+
+    // Test suite: https://regexr.com/7hoah
+    private static final Pattern MSG_JOINED_GUILD = Pattern.compile("§3You have joined §b([a-zA-Z ]*)§3!");
+
+    // Test suite: https://regexr.com/7hoak
+    private static final Pattern MSG_RANK_CHANGED = Pattern.compile(
+            "'s guild rank from (?:Recruiter|Recruit|Captain|Strategist|Chief|Owner) to (Recruiter|Recruit|Captain|Strategist|Chief|Owner)");
+
+    private String guildName = "";
+    private GuildRank guildRank;
 
     public GuildModel(CharacterModel characterModel) {
         super(List.of(characterModel));
     }
 
+    // This needs to run before any chat modifications (eg. chat mentions, filter, etc)
+    // Side note; it is currently impossible to detect when we get kicked as there are no messages sent at all
+    @SubscribeEvent(priority = EventPriority.HIGHEST)
+    public void onChatMessage(ChatMessageReceivedEvent e) {
+        StyledText message = StyledText.fromComponent(e.getMessage());
+
+        if (message.matches(MSG_LEFT_GUILD)) {
+            guildName = "";
+            guildRank = null;
+            WynntilsMod.info("User left guild");
+            return;
+        }
+
+        Matcher joinedGuildMatcher = message.getMatcher(MSG_JOINED_GUILD);
+        if (joinedGuildMatcher.matches()) {
+            guildName = joinedGuildMatcher.group(1);
+            guildRank = GuildRank.RECRUIT;
+            WynntilsMod.info("User joined guild " + guildName + " as a " + guildRank);
+            return;
+        }
+
+        if (message.startsWith("§3[INFO]§b ") && message.contains(McUtils.playerName() + "'s guild rank from")) {
```

Comment:
What about user names that end on `s`? Sometime Wynn does like `mag_icus's` and sometimes `mag_icus'`.  I can probably help you test, but I'd need bolyai to help with changing my rank.

Reply:
I am not sure. If you two could try then let me know, I would highly appreciate it

Reply:
I can help, if we organize something today or tomorrow

Reply:
![image](https://github.com/Wynntils/Artemis/assets/49001742/667f93dd-8feb-49ed-8ca6-62e1ba423eb8)